pkg/virtctl/usbredir/client.go

```diff
@@ -0,0 +1,196 @@
+//go:build !s390x
+
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright the KubeVirt Authors.
+ *
+ */
+
+package usbredir
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"os/exec"
+	"time"
+
+	kvcorev1 "kubevirt.io/client-go/generated/kubevirt/clientset/versioned/typed/core/v1"
+	"kubevirt.io/client-go/log"
+)
+
+type ClientConnectFn func(ctx context.Context, device, address string) error
+
+type Client struct {
+	// To connect local USB device buffer to the remote VM using the websocket.
+	inputReader  *io.PipeReader
+	inputWriter  *io.PipeWriter
+	outputReader *io.PipeReader
+	outputWriter *io.PipeWriter
+
+	listener *net.TCPListener
+
+	// channels
+	done   chan struct{}
+	stream chan error
+	local  chan error
+	remote chan error
+
+	ctx context.Context
+
+	ClientConnect ClientConnectFn
+}
+
+func NewUSBRedirClient(ctx context.Context, address string, stream kvcorev1.StreamInterface) (*Client, error) {
+	inReader, inWriter := io.Pipe()
+	outReader, outWriter := io.Pipe()
+	k := &Client{
+		ctx:           ctx,
+		inputReader:   inReader,
+		inputWriter:   inWriter,
+		outputReader:  outReader,
+		outputWriter:  outWriter,
+		ClientConnect: clientConnect,
+	}
+
+	// Create local TCP server for usbredir client to connect
+	if err := k.withLocalTCPClient(address); err != nil {
+		return nil, err
+	}
+
+	// Start stream with remote usbredir endpoint
+	k.withRemoteVMIStream(stream)
```

Comment:
nit: Why not to have this as a go routing instead of embedding the go routine in the function?

Reply:
This is a matter of preference, either way is fine for me. Prior iteration, this was exposed but now it is local usage only. I can change if you think is best, the only other alteration is the creation of `k.stream` channel, I'll need to move it out.